### DISCUSSION TITLE
Change objects' functions to read-only properties

### DISCUSSION THREAD

**Discussion by kqv1** [RomanArzumanyan/VALI#84] [Ideas]
First of all, thank @RomanArzumanyan for creating a great library.

Currently, most properties of various classes (ie **PyDecoder**, **PySurface**) are implemented as retrieve functions (ie **Duration()**, **Height()**, **Format()**), not as read-only properties (ie **Duration**, **Height**, **Format**)

Although they are working perfectly, it's a bit difficult to read (with the parentheses), and difficult to debug within IDE (which usuall show propeties automatically)

This can be implemented easily by replace all **.def("** to **.def_property_readonly("** in pybind11 syntax
```python
      .def_property_readonly("Duration", &PyDecoder::Duration,
           R"pbdoc(
        Return video duration time in seconds. May not be present.
    )pbdoc")

```
This is very minor change, but I think it could help new developers to catch up with this great library. The effort seems to low. I just have done the replace all in PyDecoder.cpp and the code compiled fine. 

![image](https://github.com/user-attachments/assets/a929461b-7224-4402-899b-197d208d775f)
VSCODE automatically show properties

**Comment by RomanArzumanyan** [comment]
Hi @kqv1 

Thank you for this proposal. Could you submit a PR ?
I'm currently knee deep into https://github.com/RomanArzumanyan/VALI/tree/python_vali which aims to upload the wheel to pypi.

**Reply by kqv1** [reply]
I am updating the code. Most of the changes are straightforward, but there are some cosmetic adjustments I would like you to consider:

1. Remove **Get** from some property names: **GetMotionVectors**, **GetRawMemSize**, **GetNumElems**, **GetElemSize**, **GetFrameSizeInBytes**. This way, property names are nouns, and method names are verbs.
2. Add **Is** to some bool property names: **IsAccelerated**, **IsEmpty**, **IsOwnMemory**
3. In the **PySurface** class, the **Width**, **Height**, and **Pitch** functions now accept an optional argument for the plane, with the default being 0. For simplicity, I propose making these simple properties that always return the value of plane 0. If there is a need to access properties of other planes, we can use **surface.Planes[i].Width** (see the next point).
4. Change the **PySurface.PlanePtr(i)** function to **Planes** property, which would return all planes as a tuple.

If you agree with these proposals, I will update the test cases accordingly.

Below is the list of changes:
![image](https://github.com/user-attachments/assets/4706b4cf-374b-442a-8a51-34d06bb75c86)

**Reply by RomanArzumanyan** [reply]
Hi @kqv1 

It's a bit difficult for me to grasp the changelist from your description. Could you submit a PR ?
GitHub will automatically work out the list of proposed changes.

WRT one particular proposal:

> In the PySurface class, the Width, Height, and Pitch functions now accept an optional argument for the plane, with the default being 0

I don't see any benefits in doing that because it will access lower-level `SurfacePlane` API from higher-level `Surface` API. 
Also it has a potential caveat WRT interleaved and semi-planar formats.

```python
# Not great but OK. 
# Gives you a chance to write in any language as in C.
u_width = surf_yuv420.PlanePtr(1).Width()

# You propose this, makes total sense.
# It takes away a chance to pass garbage plane index from user.
for plane in surf_yuv420.Planes():
    print(plane.Width())

# Contradicts your previous good idea by giving back user a chance to pass garbage plane index.
# Also requires user to understand the concept of planar, semi-planar and interleaved images.
# They didn't have to do that before. 
# They only mess with image planes intentionally by calling PlanePtr() or iterating through list of planes.
u_width = surf_yuv420.Width(1)

# For 1920x1080 RGB24, what value will be returned ?
# Surface width in pixels which is 1920 ? SurfacePlane raw width in elements which is 1920x3 ?
width = surf_rgb24.Width()

# Same stuff. 1080 or 1080x3/2 ?
height = surf_nv12.Height()

# More boilerplate code to handle out-of-range indices
height = surf_nv12.Height(3)
```

**Reply by kqv1** [reply]
Sorry for the unclear explanation. I have submitted a Pull Request.

**Reply by RomanArzumanyan** [reply]
Hi @kqv1 

It looks like your PR was incomplete because it didn't introduce any changes to tests. 
So I was under impression that it doesn't break existing API, but it actually does. 

I think it's a good PR because it significantly improves the debugging process, but due to major breaking API changes I'll have to merge it to `main` branch later under a major version bump. 

Currently I'm working on the pypi publishing, and I'll have to introduce some breaking changes for that. IMO that will be a good time to merge this into `main` branch.

There are no actions required on your side, I'll do the rest a bit later.

**Reply by kqv1** [reply]
Totally agreed with your point.
Thank you for considering my proposal

**Comment by RomanArzumanyan** [comment]
@kqv1 

Your MR is merged into v.4.0.0 `main`, closing the discussion as complete.
